src/cpp/session/prefs/UserStateLayer.cpp

```diff
@@ -33,18 +33,34 @@ UserStateLayer::UserStateLayer():
 
 core::Error UserStateLayer::readPrefs()
 {
-   prefsFile_ = core::system::xdg::userDataDir().completePath(
+   FilePath schemaFile = options().rResourcesPath().completePath("schema").completePath(kUserStateSchemaFile);
+
+   // desktop and server versions of RStudio use separate state files so that mixing desktop and server
+   // on the same machine is possible w/o side effects like sharing a source database
+   stateFile_ = core::system::xdg::userDataDir().completePath(
          options().programMode() == kSessionProgramModeDesktop ? 
             kUserStateFileDesktop : 
             kUserStateFileServer);
 
-   return loadPrefsFromFile(prefsFile_,
-      options().rResourcesPath().completePath("schema").completePath(kUserStateSchemaFile));
+   if (!stateFile_.exists())
+   {
+      // if there's no state file yet, check for a state file left by an older version of RStudio 1.3
+      FilePath oldStateFile = core::system::xdg::userDataDir().completePath("rstudio-state.json");
+      if (oldStateFile.exists())
+      {
+          // found an old file; attempt to migrate it
+          Error error = oldStateFile.move(stateFile_);
```

Comment:
One worry: If I understand correctly, some users might lose their old tabs if they had been working with both RStudio Desktop + RStudio Server on the same machine (since previously they "shared" the same set of tabs). Is this something worth guarding against?

Reply:
There is no practical way to guard against this since they need to use separate sets of tabs (they can't _both_ use your old tab sets; see https://github.com/rstudio/rstudio/pull/6669), so we pick a winner here (whichever you launch first) and the loser gets a new set of tabs.

There is an *impractical* solution here. We could assign new context IDs for desktop and server, remember the old ones, and then whenever we open a project, we clone the source database from the old id to the new one if we find it. However the volume of fragile, one-time-use code involved there is probably not worth it given that running desktop and server against the same account is very rare in the wild.